File: 0213-house-robber-ii/0213-house-robber-ii.java
class Solution {
    public int rob(int[] nums) {
        int n = nums.length;
        if(n == 1) {
            return nums[0];
        }
        if(n == 2) {
            return Math.max(nums[0], nums[1]);
        }
        int a = nums[0];
        int b1 = Math.max(nums[0], nums[1]);

        for(int i = 2; i < n - 1; i++) {
            int result = Math.max(b1, a + nums[i]);
            a = b1;
            b1 = result;
        }

        a = nums[1];
        int b2 = Math.max(nums[1], nums[2]);

        for(int i = 3; i < n; i++) {
            int result = Math.max(b2, a + nums[i]);
            a = b2;
            b2 = result;
        }

        return Math.max(b1, b2);
    }
}

File: 0213-house-robber-ii/0213-house-robber-ii.cpp
class Solution {
public:
    int helper(vector<int>& nums, int st, int end) {
        int n = nums.size();
        // vector<int> dp(n - 1);

        int prev1 = nums[st];
        int prev2 = max(nums[st], nums[st+1]);

        for(int i = st + 2; i <= end; i++) {
            int result = max(prev2, prev1 + nums[i]);
            prev1 = prev2;
            prev2 = result;
        }

        return prev2;
    }

    int rob(vector<int>& nums) {
        int n = nums.size();
        if(n == 1) {
            return nums[0];
        }
        if(n == 2) {
            return max(nums[0], nums[1]);
        }
        return max(helper(nums, 0, n - 2), helper(nums, 1, n - 1));
    }
};